### PR TITLE
Fix torch::nn::init::orthogonal_ with CNNs

### LINKED ISSUE #18518: "shape '[32, 3]' is invalid for input of size 6144" when initializing Conv2d layer in C++ frontend
## 🐛 Bug

When using `torch::nn::init::orthogonal_()` on a `torch::nn::Conv2d` layer's weights parameter, it fails with the following message:
```
shape '[32, 3]' is invalid for input of size 6144
```

In this case, my convolution layer has a shape of `[32, 3, 8, 8]`, which when multiplied out is 6144 weights.

Using the same initialization function on a `Linear` or `GRU` layer works fine.

## To Reproduce
```c++
auto conv_layer = torch::nn::Conv2d(nn::Conv2dOptions(3, 32, 8).stride(4));
torch::nn::init::orthogonal_(conv_layer->named_parameters()["weight"]);
```

## Expected behavior

In the Python frontend it correctly initializes the parameter, so I would expect that to happen.

## Environment

 - PyTorch Version (e.g., 1.0): 1.0.1
 - OS (e.g., Linux): Ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): ccmake .. (choose options for building libtorch), make -j8, sudo make install
 - Python version: 3.6
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: GTX 1050Ti
 - Any other relevant information: Using the PyTorch C++ frontend

### PR DESCRIPTION
**Pull request by Omegastick**
Fixes #18518

I changed the C++ API torch::nn::init::orthogonal_ implementation to match the Python implementation.

### CODE REVIEW

**Comment by Omegastick** [comment]
I haven't touched any Python files, so I assume the Python lint checks were already failing before this commit.

File: torch/csrc/api/src/nn/init.cpp [hunk 1]
```diff
@@ -123,7 +123,7 @@ Tensor orthogonal_(Tensor tensor, double gain) {
       "Only tensors with 2 or more dimensions are supported");
 
   const auto rows = tensor.size(0);
-  const auto columns = tensor.size(1);
+  const auto columns = tensor[0].numel();
```
Comment (vishwakftw) [inline torch/csrc/api/src/nn/init.cpp]: I'm a bit doubtful about the change. They're semantically the same, but the previous version is faster.

Comment (Omegastick) [reply torch/csrc/api/src/nn/init.cpp]: They're only the same for 2 dimensional tensors. For tensors of 3 dimensions or more `numel()` counts all the elements in the sub-dimensions, whereas `size(1)` only counts the size of the 1st dimension.

For example: Calling `size(1)` on a [3, 4, 5, 6] shaped tensor returns 4, but calling `tensor[0].numel()` returns 120.

Comment (vishwakftw) [reply torch/csrc/api/src/nn/init.cpp]: Looking at the code below, this is what I think is correct:
```
const auto rows = tensor.size(-2);
const auto columns = tensor.size(-1);
```

A matrix of size (rows x columns) is created whose QR decomposition is performed and then is copied to the input tensor. We might have to expand and copy the tensor for tensors of dim > 2.

Comment (Omegastick) [reply torch/csrc/api/src/nn/init.cpp]: Yeah, that would avoid an unnecessary potential copy. I can change it to do that instead. Should we change the [Python implementation](https://github.com/pytorch/pytorch/blob/90b8552c9811ee6e9bbc9dff1f1df30c00b663dd/torch/nn/init.py#L348) too?

Comment (vishwakftw) [reply torch/csrc/api/src/nn/init.cpp]: Yeah, that'd be great.

**Comment by ezyang** [comment]
@Omegastick Thank you. Can we get a little test for this case?

@pytorchbot rebase this please

**Comment by Omegastick** [comment]
@ezyang Right, how's that? Let me know if I did something wrong with the test, I'm not too familiar with PyTorch's tests.

File: torch/nn/init.py [hunk 2]
```diff
@@ -345,7 +345,7 @@ def orthogonal_(tensor, gain=1):
         raise ValueError("Only tensors with 2 or more dimensions are supported")
 
     rows = tensor.size(0)
-    cols = tensor.numel() / rows
+    cols = int(tensor.numel() / rows)
```
Comment (vishwakftw) [inline torch/nn/init.py]: This can be made `tensor.numel() // rows`.

Comment (Omegastick) [reply torch/nn/init.py]: I amended the last commit.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**